# Kansas City Postal Code Clustering
Created by: Brandon Lombardino  
Created Date: 05/12/2020

## Introduction

A business wants to know which areas of Kansas City are ideal to open a coffee shop.  The owner hopes to draw in customers by building a unique relationship within the community rather than through low prices or innovative products.  The report will investigate which area in Kansas City is best suited for the business given the owners goals and expectations

## Data and Sources

The data is based on postal codes from the Kansas City area.  The postal codes are used to define regions within the Kansas City metropolitan area and have coordinates to map their locations.  The data will be used with folium maps and a Scikit-Learn cluster model to segment the region based on venue information imported from Foursquare.

## Methodology

### Import Libraries

In [285]:
import numpy as np
import pandas as pd
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)
import json
from geopy.geocoders import Nominatim
import requests
from pandas.io.json import json_normalize
import matplotlib.cm as cm
import matplotlib.colors as colors
from sklearn.cluster import KMeans
import folium
import geocoder

print('Libraries imported.')

Libraries imported.


### Import CSV

In [286]:
df_loc = pd.read_csv(r'C:\Users\brand\OneDrive\Documents\Master\IBM Data Science\Final Project\kc_loc.csv')
df_loc.head()

postal_code   latitude  longitude
0        64102  39.092001 -94.601628
1        64105  39.104837 -94.594928
2        64106  39.106210 -94.573278
3        64108  39.085612 -94.584134
4        64109  39.066060 -94.567489

### Check the Number of Postal Codes within Kansas City

In [287]:
print('The dataframe has {} postal codes.'.format(
        len(df_loc['postal_code'].unique()),
        df_loc.shape[0]
    )
)

The dataframe has 70 postal codes.


### Find the Kansas City Coordinates

In [288]:
address = 'Kansas City, MO'

geolocator = Nominatim(user_agent="to_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinates of Kansas City are {}, {}.'.format(latitude, longitude))

The geograpical coordinates of Kansas City are 39.100105, -94.5781416.


### Create a Map of Kansas City

In [289]:
map_kc = folium.Map(location=[latitude, longitude], zoom_start=10.4)

# add markers to map
for lat, lng, postal_code in zip(df_loc['latitude'],df_loc['longitude'], df_loc['postal_code']):
    label = '{}'.format(postal_code)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=3,
        popup=label,
        color='#E42222',
        fill=True,
        fill_color='#CB9D5B',
        fill_opacity=0.7,
        parse_html=False).add_to(map_kc)  
    
map_kc

### Use Foursquare to Explore the Area

In [290]:
CLIENT_ID = 'HRH4TIEJJO2CJFGGCCKRNAKWCMGRGQDXYCJPBHAEISQTMUQ0'
CLIENT_SECRET = 'LVEZOSHHBIYLKQODSBI5OLB01UOL3XPY4JDMUHUFEK3NDLZN'
VERSION = '20200512'

### Find the Top 500 Venues within a 1,000 Meter Radius

In [291]:
LIMIT = 500 # limit of number of venues returned by Foursquare API
radius = 1000 # define radius
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    latitude, 
    longitude, 
    radius, 
    LIMIT)
url

'https://api.foursquare.com/v2/venues/explore?&client_id=HRH4TIEJJO2CJFGGCCKRNAKWCMGRGQDXYCJPBHAEISQTMUQ0&client_secret=LVEZOSHHBIYLKQODSBI5OLB01UOL3XPY4JDMUHUFEK3NDLZN&v=20200512&ll=39.100105,-94.5781416&radius=1000&limit=500'

### Extract the Venue's Category Type

In [292]:
results = requests.get(url).json()

def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

### Add the Venue and Category Columns to the Postal Code Table

In [293]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues)

filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

KeyError: 'groups'

### Create Venue Related Data

In [261]:
def getNearbyVenues(names, latitudes, longitudes, radius=1000):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['postal_code', 
                  'latitude', 
                  'longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)
kc_venues = getNearbyVenues(names=df_loc['postal_code'],
                                   latitudes=df_loc['latitude'],
                                   longitudes=df_loc['longitude']
                                  )

64102
64105
64106
64108
64109
64101
64110
64111


KeyError: 'groups'

### Check the Structure of the Display Table

In [262]:
print(kc_venues.shape)
kc_venues.head()

(2309, 8)


index  postal_code   latitude  longitude                   Venue  \
0      0        64102  39.092001 -94.601628                Voltaire   
1      1        64102  39.092001 -94.601628               Pole Worx   
2      2        64102  39.092001 -94.601628  Stockyards Brewing Co.   
3      3        64102  39.092001 -94.601628      The Westside Local   
4      4        64102  39.092001 -94.601628               Chez Elle   

   Venue Latitude  Venue Longitude       Venue Category  
0       39.093935       -94.604779  American Restaurant  
1       39.094161       -94.604701         Dance Studio  
2       39.094241       -94.605101              Brewery  
3       39.093300       -94.593200            Gastropub  
4       39.092578       -94.593221    French Restaurant

### Group the Number of Venues by Postal Code

In [263]:
kc_venues_groupby = kc_venues.groupby('postal_code').count()

kc_venues_groupby

index  latitude  longitude  Venue  Venue Latitude  \
postal_code                                                      
64101           37        37         37     37              37   
64102           38        38         38     38              38   
64105           56        56         56     56              56   
64106           69        69         69     69              69   
64108          100       100        100    100             100   
64109           13        13         13     13              13   
64110           27        27         27     27              27   
64111           53        53         53     53              53   
64112           83        83         83     83              83   
64113           36        36         36     36              36   
64116           42        42         42     42              42   
64118           30        30         30     30              30   
64121           10        10         10     10              10   
64124           20        20         20     20              20   
64127           13        13         13     13              13   
64130           14        14         14     14              14   
64141          100       100        100    100             100   
64144          100       100        100    100             100   
64145           54        54         54     54              54   
64148          100       100        100    100             100   
64151           13        13         13     13              13   
64168           10        10         10     10              10   
64170          100       100        100    100             100   
64171           32        32         32     32              32   
64179          100       100        100    100             100   
64180          100       100        100    100             100   
64184          100       100        100    100             100   
64187          100       100        100    100             100   
64188           32        32         32     32              32   
64190          100       100        100    100             100   
64191          100       100        100    100             100   
64195           27        27         27     27              27   
64196          100       100        100    100             100   
64197          100       100        100    100             100   
64198          100       100        100    100             100   
64199          100       100        100    100             100   
64999          100       100        100    100             100   

             Venue Longitude  Venue Category  
postal_code                                   
64101                     37              37  
64102                     38              38  
64105                     56              56  
64106                     69              69  
64108                    100             100  
64109                     13              13  
64110                     27              27  
64111                     53              53  
64112                     83              83  
64113                     36              36  
64116                     42              42  
64118                     30              30  
64121                     10              10  
64124                     20              20  
64127                     13              13  
64130                     14              14  
64141                    100             100  
64144                    100             100  
64145                     54              54  
64148                    100             100  
64151                     13              13  
64168                     10              10  
64170                    100             100  
64171                     32              32  
64179                    100             100  
64180                    100             100  
64184                    100             100  
64187                    100             100  
64188                     32

In [264]:
under_ten = kc_venues_groupby[kc_venues_groupby["Venue"] < 10]

under_ten

Empty DataFrame
Columns: [index, latitude, longitude, Venue, Venue Latitude, Venue Longitude, Venue Category]
Index: []

### Drop Postal Codes that have Less than 10 Venues

In [265]:
kc_venues_1 = kc_venues[kc_venues["postal_code"] != 64114]
kc_venues_2 = kc_venues_1[kc_venues["postal_code"] != 64117]
kc_venues_3 = kc_venues_2[kc_venues["postal_code"] != 64119]
kc_venues_4 = kc_venues_3[kc_venues["postal_code"] != 64120]
kc_venues_5 = kc_venues_4[kc_venues["postal_code"] != 64123]
kc_venues_6 = kc_venues_5[kc_venues["postal_code"] != 64125]
kc_venues_7 = kc_venues_6[kc_venues["postal_code"] != 64126]
kc_venues_8 = kc_venues_7[kc_venues["postal_code"] != 64128]
kc_venues_9 = kc_venues_8[kc_venues["postal_code"] != 64129]
kc_venues_10 = kc_venues_9[kc_venues["postal_code"] != 64131]
kc_venues_11 = kc_venues_10[kc_venues["postal_code"] != 64132]
kc_venues_12 = kc_venues_11[kc_venues["postal_code"] != 64133]
kc_venues_13 = kc_venues_12[kc_venues["postal_code"] != 64134]
kc_venues_14 = kc_venues_13[kc_venues["postal_code"] != 64136]
kc_venues_15 = kc_venues_14[kc_venues["postal_code"] != 64137]
kc_venues_16 = kc_venues_15[kc_venues["postal_code"] != 64138]
kc_venues_17 = kc_venues_16[kc_venues["postal_code"] != 64139]
kc_venues_18 = kc_venues_17[kc_venues["postal_code"] != 64146]
kc_venues_19 = kc_venues_18[kc_venues["postal_code"] != 64147]
kc_venues_20 = kc_venues_19[kc_venues["postal_code"] != 64152]
kc_venues_21 = kc_venues_20[kc_venues["postal_code"] != 64153]
kc_venues_22 = kc_venues_21[kc_venues["postal_code"] != 64154]
kc_venues_23 = kc_venues_22[kc_venues["postal_code"] != 64155]
kc_venues_24 = kc_venues_23[kc_venues["postal_code"] != 64156]
kc_venues_25 = kc_venues_24[kc_venues["postal_code"] != 64157]
kc_venues_26 = kc_venues_25[kc_venues["postal_code"] != 64158]
kc_venues_27 = kc_venues_26[kc_venues["postal_code"] != 64161]
kc_venues = kc_venues_27[kc_venues["postal_code"] != 64163]
kc_venues.reset_index(inplace = True)

print(kc_venues.shape)
kc_venues

(2309, 9)


level_0  index  postal_code   latitude  longitude  \
0           0      0        64102  39.092001 -94.601628   
1           1      1        64102  39.092001 -94.601628   
2           2      2        64102  39.092001 -94.601628   
3           3      3        64102  39.092001 -94.601628   
4           4      4        64102  39.092001 -94.601628   
5           5      5        64102  39.092001 -94.601628   
6           6      6        64102  39.092001 -94.601628   
7           7      7        64102  39.092001 -94.601628   
8           8      8        64102  39.092001 -94.601628   
9           9      9        64102  39.092001 -94.601628   
10         10     10        64102  39.092001 -94.601628   
11         11     11        64102  39.092001 -94.601628   
12         12     12        64102  39.092001 -94.601628   
13         13     13        64102  39.092001 -94.601628   
14         14     14        64102  39.092001 -94.601628   
15         15     15        64102  39.092001 -94.601628   
16         16     16        64102  39.092001 -94.601628   
17         17     17        64102  39.092001 -94.601628   
18         18     18        64102  39.092001 -94.601628   
19         19     19        64102  39.092001 -94.601628   
20         20     20        64102  39.092001 -94.601628   
21         21     21        64102  39.092001 -94.601628   
22         22     22        64102  39.092001 -94.601628   
23         23     23        64102  39.092001 -94.601628   
24         24     24        64102  39.092001 -94.601628   
25         25     25        64102  39.092001 -94.601628   
26         26     26        64102  39.092001 -94.601628   
27         27     27        64102  39.092001 -94.601628   
28         28     28        64102  39.092001 -94.601628   
29         29     29        64102  39.092001 -94.601628   
30         30     30        64102  39.092001 -94.601628   
31         31     31        64102  39.092001 -94.601628   
32         32     32        64102  39.092001 -94.601628   
33         33     33        64102  39.092001 -94.601628   
34         34     34        64102  39.092001 -94.601628   
35         35     35        64102  39.092001 -94.601628   
36         36     36        64102  39.092001 -94.601628   
37         37     37        64102  39.092001 -94.601628   
38         38     38        64105  39.104837 -94.594928   
39         39     39        64105  39.104837 -94.594928   
40         40     40        64105  39.104837 -94.594928   
41         41     41        64105  39.104837 -94.594928   
42         42     42        64105  39.104837 -94.594928   
43         43     43        64105  39.104837 -94.594928   
44         44     44        64105  39.104837 -94.594928   
45         45     45        64105  39.104837 -94.594928   
46         46     46        64105  39.104837 -94.594928   
47         47     47        64105  39.104837 -94.594928   
48         48     48        64105  39.104837 -94.594928   
49         49     49        64105  39.104837 -94.594928   
50         50     50        64105  39.104837 -94.594928   
51         51     51        64105  39.104837 -94.594928   
52         52     52        64105  39.104837 -94.594928   
53         53     53        64105  39.104837 -94.594928   
54         54     54        64105  39.104837 -94.594928   
55         55     55        64105  39.104837 -94.594928   
56         56     56        64105  39.104837 -94.594928   
57         57     57        64105  39.104837 -94.594928   
58         58     58        64105  39.104837 -94.594928   
59         59     59        64105  39.104837 -94.594928   
60         60     60        64105  39.104837 -94.594928   
61         61     61        64105  39.104837 -94.594928   
62         62     62        64105  39.104837 -94.594928   
63         63     63        64105  39.104837 -94.594928   
64         64     64        64105  39.104837 -94.594928   
65         65     65        64105  39.104837 -94.594928   
66         66     66        64105  39.104837 -94.5949

### Use One Hot Encoding on the Venue Table

In [266]:
kc_onehot = pd.get_dummies(kc_venues[['Venue Category']], prefix="", prefix_sep="")
kc_onehot['postal_code'] = kc_venues['postal_code']
fixed_columns = [kc_onehot.columns[-1]] + list(kc_onehot.columns[:-1])
kc_onehot = kc_onehot[fixed_columns]

kc_onehot.head()

postal_code  Accessories Store  Airport  Airport Service  Airport Terminal  \
0        64102                  0        0                0                 0   
1        64102                  0        0                0                 0   
2        64102                  0        0                0                 0   
3        64102                  0        0                0                 0   
4        64102                  0        0                0                 0   

   American Restaurant  Antique Shop  Aquarium  Arcade  Art Gallery  \
0                    1             0         0       0            0   
1                    0             0         0       0            0   
2                    0             0         0       0            0   
3                    0             0         0       0            0   
4                    0             0         0       0            0   

   Arts & Crafts Store  Athletics & Sports  Auto Dealership  Automotive Shop  \
0                    0                   0                0                0   
1                    0                   0                0                0   
2                    0                   0                0                0   
3                    0                   0                0                0   
4                    0                   0                0                0   

   BBQ Joint  Bagel Shop  Bakery  Bank  Bar  Baseball Field  Bed & Breakfast  \
0          0           0       0     0    0               0                0   
1          0           0       0     0    0               0                0   
2          0           0       0     0    0               0                0   
3          0           0       0     0    0               0                0   
4          0           0       0     0    0               0                0   

   Beer Bar  Beer Garden  Big Box Store  Bike Rental / Bike Share  Bistro  \
0         0            0              0                         0       0   
1         0            0              0                         0       0   
2         0            0              0                         0       0   
3         0            0              0                         0       0   
4         0            0              0                         0       0   

   Board Shop  Boat or Ferry  Bookstore  Border Crossing  Boutique  \
0           0              0          0                0         0   
1           0              0          0                0         0   
2           0              0          0                0         0   
3           0              0          0                0         0   
4           0              0          0                0         0   

   Brazilian Restaurant  Breakfast Spot  Brewery  Bridge  Bubble Tea Shop  \
0                     0               0        0       0                0   
1                     0               0        0       0                0   
2                     0               0        1       0                0   
3                     0               0        0       0                0   
4                     0               0        0       0                0   

   Building  Burger Joint  Burrito Place  Bus Line  Bus Station  Bus Stop  \
0         0             0              0         0            0         0   
1         0             0              0         0            0         0   
2         0             0              0         0            0         0   
3         0             0              0         0            0         0   
4         0             0              0         0            0         0   

   Business Service  Butcher  Café  Cajun / Creole Restaurant  Candy Store  \
0                 0        0     0                          0            0   
1                 0        0     0                          0            0   
2                 0        0     0                          0            0   
3    

### Group Rows by the Postal Code and by Taking the Mean of the Frequency of Occurance of Each Category

In [267]:
kc_grouped = kc_onehot.groupby('postal_code').mean().reset_index()

kc_grouped

postal_code  Accessories Store   Airport  Airport Service  \
0         64101           0.000000  0.000000         0.000000   
1         64102           0.000000  0.000000         0.000000   
2         64105           0.000000  0.000000         0.000000   
3         64106           0.000000  0.000000         0.000000   
4         64108           0.000000  0.000000         0.000000   
5         64109           0.000000  0.000000         0.000000   
6         64110           0.000000  0.000000         0.000000   
7         64111           0.000000  0.000000         0.000000   
8         64112           0.012048  0.000000         0.000000   
9         64113           0.000000  0.000000         0.000000   
10        64116           0.000000  0.000000         0.000000   
11        64118           0.000000  0.000000         0.000000   
12        64121           0.000000  0.000000         0.000000   
13        64124           0.000000  0.000000         0.000000   
14        64127           0.000000  0.000000         0.000000   
15        64130           0.000000  0.000000         0.000000   
16        64141           0.000000  0.000000         0.000000   
17        64144           0.000000  0.000000         0.000000   
18        64145           0.000000  0.000000         0.000000   
19        64148           0.000000  0.000000         0.000000   
20        64151           0.000000  0.000000         0.000000   
21        64168           0.000000  0.000000         0.000000   
22        64170           0.000000  0.000000         0.000000   
23        64171           0.000000  0.000000         0.000000   
24        64179           0.000000  0.000000         0.000000   
25        64180           0.000000  0.000000         0.000000   
26        64184           0.000000  0.000000         0.000000   
27        64187           0.000000  0.000000         0.000000   
28        64188           0.000000  0.000000         0.000000   
29        64190           0.000000  0.000000         0.000000   
30        64191           0.000000  0.000000         0.000000   
31        64195           0.000000  0.037037         0.037037   
32        64196           0.000000  0.000000         0.000000   
33        64197           0.000000  0.000000         0.000000   
34        64198           0.000000  0.000000         0.000000   
35        64199           0.000000  0.000000         0.000000   
36        64999           0.000000  0.000000         0.000000   

    Airport Terminal  American Restaurant  Antique Shop  Aquarium    Arcade  \
0           0.000000             0.027027      0.162162      0.00  0.000000   
1           0.000000             0.026316      0.078947      0.00  0.000000   
2           0.000000             0.017857      0.125000      0.00  0.000000   
3           0.000000             0.028986      0.000000      0.00  0.000000   
4           0.000000             0.050000      0.000000      0.00  0.010000   
5           0.000000             0.076923      0.000000      0.00  0.000000   
6           0.000000             0.000000      0.000000      0.00  0.000000   
7           0.000000             0.018868      0.000000      0.00  0.000000   
8           0.000000             0.084337      0.000000      0.00  0.000000   
9           0.000000             0.055556      0.000000      0.00  0.000000   
10          0.000000             0.023810      0.000000      0.00  0.000000   
11          0.000000             0.066667      0.000000      0.00  0.033333   
12          0.000000             0.000000      0.000000      0.00  0.000000   
13          0.000000             0.000000      0.000000      0.00  0.000000   
14          0.000000             0.076923      0.000000      0.00  0.000000   
15          0.000000             0.071429      0.000000      0.00  0.000000   
16          0.000000             0.040000      0.000000      0.00  0.010000   
17          0.000000             0.050000      0.000000      0.01  0.010000   
18          0.000000            

### Display Each Postal Code with Top 5 Most Common Venues

In [268]:
num_top_venues = 5

for code in kc_grouped['postal_code']:
    print("----"+str(code)+"----")
    temp = kc_grouped[kc_grouped['postal_code'] == code].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----64101----
             venue  freq
0     Antique Shop  0.16
1             Park  0.08
2              Gym  0.05
3  Border Crossing  0.05
4      Coffee Shop  0.03


----64102----
               venue  freq
0       Antique Shop  0.08
1        Art Gallery  0.08
2                Gym  0.05
3  French Restaurant  0.05
4               Park  0.05


----64105----
           venue  freq
0   Antique Shop  0.12
1       Dive Bar  0.05
2     Food Truck  0.04
3  Deli / Bodega  0.04
4        Theater  0.04


----64106----
                   venue  freq
0         Sandwich Place  0.06
1  Vietnamese Restaurant  0.06
2     Italian Restaurant  0.04
3               Dive Bar  0.04
4                   Park  0.04


----64108----
                     venue  freq
0      American Restaurant  0.05
1  New American Restaurant  0.05
2                  Brewery  0.04
3       Mexican Restaurant  0.04
4              Coffee Shop  0.04


----64109----
                    venue  freq
0           Deli / Bodega  0.08
1  Thrif

### Place the Top 10 Most Common Venues into a Pandas Dataframe

In [269]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['postal_code']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
postal_code_venues_sorted = pd.DataFrame(columns=columns)
postal_code_venues_sorted['postal_code'] = kc_grouped['postal_code']

for ind in np.arange(kc_grouped.shape[0]):
    postal_code_venues_sorted.iloc[ind, 1:] = return_most_common_venues(kc_grouped.iloc[ind, :], num_top_venues)

postal_code_venues_sorted.head()

postal_code    1st Most Common Venue 2nd Most Common Venue  \
0        64101             Antique Shop                  Park   
1        64102             Antique Shop           Art Gallery   
2        64105             Antique Shop              Dive Bar   
3        64106    Vietnamese Restaurant        Sandwich Place   
4        64108  New American Restaurant   American Restaurant   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0       Border Crossing                   Gym      Pedestrian Plaza   
1                   Gym     French Restaurant                  Park   
2         Deli / Bodega                   Gym               Theater   
3                  Park    Italian Restaurant           Coffee Shop   
4    Mexican Restaurant           Coffee Shop               Brewery   

             6th Most Common Venue 7th Most Common Venue  \
0  Southern / Soul Food Restaurant           Social Club   
1          New American Restaurant               Brewery   
2                            Hotel            Food Truck   
3                      Pizza Place              Dive Bar   
4                            Hotel                   Gym   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0              Building                  Café            Coffee Shop  
1    Salon / Barbershop     Recreation Center         Clothing Store  
2                  Park          Cocktail Bar            Coffee Shop  
3                Bakery     French Restaurant          Grocery Store  
4    Italian Restaurant          Burger Joint        Thai Restaurant

### Create Postal Code Clusters

In [270]:
# set number of clusters
kclusters = 5

kc_grouped_clustering = kc_grouped.drop('postal_code', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(kc_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

# add clustering labels
postal_code_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

df_kc = df_loc

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
df_kc = df_kc.join(postal_code_venues_sorted.set_index('postal_code'), on='postal_code')

df_kc.head() # check the last columns!

postal_code   latitude  longitude  Cluster Labels    1st Most Common Venue  \
0        64102  39.092001 -94.601628             2.0             Antique Shop   
1        64105  39.104837 -94.594928             2.0             Antique Shop   
2        64106  39.106210 -94.573278             0.0    Vietnamese Restaurant   
3        64108  39.085612 -94.584134             0.0  New American Restaurant   
4        64109  39.066060 -94.567489             3.0   Thrift / Vintage Store   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0           Art Gallery                   Gym     French Restaurant   
1              Dive Bar         Deli / Bodega                   Gym   
2        Sandwich Place                  Park    Italian Restaurant   
3   American Restaurant    Mexican Restaurant           Coffee Shop   
4                  Park  Fast Food Restaurant              Bus Stop   

  5th Most Common Venue    6th Most Common Venue 7th Most Common Venue  \
0                  Park  New American Restaurant               Brewery   
1               Theater                    Hotel            Food Truck   
2           Coffee Shop              Pizza Place              Dive Bar   
3               Brewery                    Hotel                   Gym   
4        Discount Store                  Theater             Juice Bar   

  8th Most Common Venue 9th Most Common Venue           10th Most Common Venue  
0    Salon / Barbershop     Recreation Center                   Clothing Store  
1                  Park          Cocktail Bar                      Coffee Shop  
2                Bakery     French Restaurant                    Grocery Store  
3    Italian Restaurant          Burger Joint                  Thai Restaurant  
4       Automotive Shop           Candy Store  Southern / Soul Food Restaurant

### Check the Cluster Label Distribution

In [271]:
df_kc.fillna(-1).groupby(["Cluster Labels"]).count()

postal_code  latitude  longitude  1st Most Common Venue  \
Cluster Labels                                                            
-1.0                     33        33         33                     33   
 0.0                     19        19         19                     19   
 1.0                      1         1          1                      1   
 2.0                      5         5          5                      5   
 3.0                     11        11         11                     11   
 4.0                      1         1          1                      1   

                2nd Most Common Venue  3rd Most Common Venue  \
Cluster Labels                                                 
-1.0                               33                     33   
 0.0                               19                     19   
 1.0                                1                      1   
 2.0                                5                      5   
 3.0                               11                     11   
 4.0                                1                      1   

                4th Most Common Venue  5th Most Common Venue  \
Cluster Labels                                                 
-1.0                               33                     33   
 0.0                               19                     19   
 1.0                                1                      1   
 2.0                                5                      5   
 3.0                               11                     11   
 4.0                                1                      1   

                6th Most Common Venue  7th Most Common Venue  \
Cluster Labels                                                 
-1.0                               33                     33   
 0.0                               19                     19   
 1.0                                1                      1   
 2.0                                5                      5   
 3.0                               11                     11   
 4.0                                1                      1   

                8th Most Common Venue  9th Most Common Venue  \
Cluster Labels                                                 
-1.0                               33                     33   
 0.0                               19                     19   
 1.0                                1                      1   
 2.0                                5                      5   
 3.0                               11                     11   
 4.0                                1                      1   

                10th Most Common Venue  
Cluster Labels                          
-1.0                                33  
 0.0                                19  
 1.0                                 1  
 2.0                                 5  
 3.0                                11  
 4.0                                 1

### Drop NAN Clusters

In [272]:
df_kc.dropna(inplace = True)

df_kc

postal_code   latitude  longitude  Cluster Labels  \
0         64102  39.092001 -94.601628             2.0   
1         64105  39.104837 -94.594928             2.0   
2         64106  39.106210 -94.573278             0.0   
3         64108  39.085612 -94.584134             0.0   
4         64109  39.066060 -94.567489             3.0   
5         64101  39.104106 -94.599688             2.0   
6         64110  39.034480 -94.572904             3.0   
7         64111  39.057304 -94.593710             0.0   
8         64112  39.036082 -94.595006             0.0   
9         64113  39.013748 -94.595464             3.0   
11        64116  39.146376 -94.570772             3.0   
17        64127  39.089378 -94.538050             3.0   
24        64118  39.213954 -94.575082             3.0   
25        64121  39.129600 -94.510700             1.0   
27        64124  39.107713 -94.539608             3.0   
28        64130  39.033640 -94.541624             3.0   
33        64141  39.099700 -94.578400             0.0   
34        64144  39.082900 -94.589100             0.0   
35        64148  39.099700 -94.578400             0.0   
38        64145  38.880850 -94.596219             3.0   
41        64151  39.212680 -94.632686             3.0   
50        64170  39.099700 -94.578400             0.0   
51        64171  39.103300 -94.600900             2.0   
52        64187  39.099700 -94.578400             0.0   
53        64188  39.103300 -94.600900             2.0   
54        64197  39.099700 -94.578400             0.0   
55        64198  39.099700 -94.578400             0.0   
56        64199  39.099700 -94.578400             0.0   
61        64168  39.177100 -94.613100             3.0   
62        64179  39.099700 -94.578400             0.0   
63        64180  39.099700 -94.578400             0.0   
64        64184  39.099700 -94.578400             0.0   
65        64190  39.081600 -94.588900             0.0   
66        64191  39.099700 -94.578400             0.0   
67        64195  39.305200 -94.715700             4.0   
68        64196  39.099700 -94.578400             0.0   
69        64999  39.099400 -94.578300             0.0   

      1st Most Common Venue 2nd Most Common Venue   3rd Most Common Venue  \
0              Antique Shop           Art Gallery                     Gym   
1              Antique Shop              Dive Bar           Deli / Bodega   
2     Vietnamese Restaurant        Sandwich Place                    Park   
3   New American Restaurant   American Restaurant      Mexican Restaurant   
4    Thrift / Vintage Store                  Park    Fast Food Restaurant   
5              Antique Shop                  Park         Border Crossing   
6            Sandwich Place   Fried Chicken Joint    Fast Food Restaurant   
7               Coffee Shop          Burger Joint            Liquor Store   
8       American Restaurant                 Hotel      Seafood Restaurant   
9             Grocery Store           Coffee Shop     American Restaurant   
11           Sandwich Place         Grocery Store    Fast Food Restaurant   
17       Mexican Restaurant        Sandwich Place          Discount Store   
24                     Park  Fast Food Restaurant     American Restaurant   
25       Italian Restaurant        Breakfast Spot       Convenience Store   
27                     Park    Mexican Restaurant       Convenience Store   
28     Fast Food Restaurant     Electronics Store          Shopping Plaza   
33                  Theater                 Hotel             Coffee Shop   
34      American Restaurant               Brewery          History Museum   
35                  Theater                 Hotel             Coffee Shop   
38           Sandwich Place         Big Box Store  Furniture / Home Store   
41              Coffee Shop           Video Store    Fast Food Restaurant   
50                  Theater                 Hotel             Coffee Shop   
51             Antique Shop       Border Crossing                 

### Change Postal Code from a Float into a String

In [273]:
df_kc = df_kc.astype({"Cluster Labels": int})
df_kc.reset_index(inplace = True)

df_kc.head()

index  postal_code   latitude  longitude  Cluster Labels  \
0      0        64102  39.092001 -94.601628               2   
1      1        64105  39.104837 -94.594928               2   
2      2        64106  39.106210 -94.573278               0   
3      3        64108  39.085612 -94.584134               0   
4      4        64109  39.066060 -94.567489               3   

     1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0             Antique Shop           Art Gallery                   Gym   
1             Antique Shop              Dive Bar         Deli / Bodega   
2    Vietnamese Restaurant        Sandwich Place                  Park   
3  New American Restaurant   American Restaurant    Mexican Restaurant   
4   Thrift / Vintage Store                  Park  Fast Food Restaurant   

  4th Most Common Venue 5th Most Common Venue    6th Most Common Venue  \
0     French Restaurant                  Park  New American Restaurant   
1                   Gym               Theater                    Hotel   
2    Italian Restaurant           Coffee Shop              Pizza Place   
3           Coffee Shop               Brewery                    Hotel   
4              Bus Stop        Discount Store                  Theater   

  7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0               Brewery    Salon / Barbershop     Recreation Center   
1            Food Truck                  Park          Cocktail Bar   
2              Dive Bar                Bakery     French Restaurant   
3                   Gym    Italian Restaurant          Burger Joint   
4             Juice Bar       Automotive Shop           Candy Store   

            10th Most Common Venue  
0                   Clothing Store  
1                      Coffee Shop  
2                    Grocery Store  
3                  Thai Restaurant  
4  Southern / Soul Food Restaurant

### Map the Clusters

In [274]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(df_kc['latitude'], df_kc['longitude'], df_kc['postal_code'], df_kc['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

## Results

### Cluster 1 - City Centre
Represented on the map in red \
Defined by different types of bars and entertainment venues \
Caters to nightlife and tourists

In [248]:
df_kc.loc[df_kc['Cluster Labels'] == 0, df_kc.columns[[1] + list(range(5, df_kc.shape[1]))]]

postal_code    1st Most Common Venue 2nd Most Common Venue  \
2         64106    Vietnamese Restaurant        Sandwich Place   
3         64108  New American Restaurant   American Restaurant   
7         64111              Coffee Shop          Burger Joint   
8         64112      American Restaurant                 Hotel   
16        64141                  Theater                 Hotel   
17        64144      American Restaurant               Brewery   
18        64148                  Theater                 Hotel   
21        64170                  Theater                 Hotel   
23        64187                  Theater                 Hotel   
25        64197                  Theater                 Hotel   
26        64198                  Theater                 Hotel   
27        64199                  Theater                 Hotel   
29        64179                  Theater                 Hotel   
30        64180                  Theater                 Hotel   
31        64184                  Theater                 Hotel   
32        64190       Mexican Restaurant               Brewery   
33        64191                  Theater                 Hotel   
35        64196                  Theater                 Hotel   
36        64999                  Theater               Brewery   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
2                   Park    Italian Restaurant           Coffee Shop   
3     Mexican Restaurant           Coffee Shop               Brewery   
7           Liquor Store              Dive Bar               Gay Bar   
8     Seafood Restaurant           Pizza Place        Cosmetics Shop   
16           Coffee Shop   American Restaurant                Lounge   
17        History Museum    Mexican Restaurant           Coffee Shop   
18           Coffee Shop   American Restaurant                Lounge   
21           Coffee Shop   American Restaurant                Lounge   
23           Coffee Shop   American Restaurant                Lounge   
25           Coffee Shop   American Restaurant                Lounge   
26           Coffee Shop   American Restaurant                Lounge   
27           Coffee Shop   American Restaurant                Lounge   
29           Coffee Shop   American Restaurant                Lounge   
30           Coffee Shop   American Restaurant                Lounge   
31           Coffee Shop   American Restaurant                Lounge   
32   American Restaurant        History Museum                Museum   
33           Coffee Shop   American Restaurant                Lounge   
35           Coffee Shop   American Restaurant                Lounge   
36           Coffee Shop                 Hotel   American Restaurant   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
2            Pizza Place              Dive Bar                Bakery   
3                  Hotel                   Gym    Italian Restaurant   
7             Smoke Shop            Print Shop          Cocktail Bar   
8            Coffee Shop        Sandwich Place    Mexican Restaurant   
16               Brewery           Music Venue            Sports Bar   
17                  Park                   Gym          Burger Joint   
18               Brewery           Music Venue            Sports Bar   
21               Brewery           Music Venue            Sports Bar   
23               Brewery           Music Venue            Sports Bar   
25               Brewery           Music Venue            Sports Bar   
26               Brewery           Music Venue            Sports Bar   
27               Brewery           Music Venue            Sports Bar   
29               Brewery           Music Venue            Sports Bar   
30               Brewery           Music Venue            Sports Bar   
31               Brewery           Music Venue            Sports Bar   
32                   Gym           Coffee Shop                 Hotel   
33               B

### Cluster 2 - Industrial Oldtown
Represented on the map in blue \
Defined by antique shops, border crossings, and parks \
Caters to the old town and blue collar community

In [249]:
df_kc.loc[df_kc['Cluster Labels'] == 2, df_kc.columns[[1] + list(range(5, df_kc.shape[1]))]]

postal_code 1st Most Common Venue 2nd Most Common Venue  \
0         64102          Antique Shop           Art Gallery   
1         64105          Antique Shop              Dive Bar   
5         64101          Antique Shop                  Park   
22        64171          Antique Shop       Border Crossing   
24        64188          Antique Shop       Border Crossing   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0                    Gym     French Restaurant                  Park   
1          Deli / Bodega                   Gym               Theater   
5        Border Crossing                   Gym      Pedestrian Plaza   
22                  Park          Cocktail Bar           Art Gallery   
24                  Park          Cocktail Bar           Art Gallery   

              6th Most Common Venue 7th Most Common Venue  \
0           New American Restaurant               Brewery   
1                             Hotel            Food Truck   
5   Southern / Soul Food Restaurant           Social Club   
22             Kitchen Supply Store           Coffee Shop   
24             Kitchen Supply Store           Coffee Shop   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0     Salon / Barbershop     Recreation Center         Clothing Store  
1                   Park          Cocktail Bar            Coffee Shop  
5               Building                  Café            Coffee Shop  
22  Gym / Fitness Center                  Café      Electronics Store  
24  Gym / Fitness Center                  Café      Electronics Store

### Cluster 3 - Suburbs
Represented on the map in green \
Defined by restaurants and shopping centers \
Caters to residents who value space and quiet over community and entertainment

In [250]:
df_kc.loc[df_kc['Cluster Labels'] == 3, df_kc.columns[[1] + list(range(5, df_kc.shape[1]))]]

postal_code   1st Most Common Venue 2nd Most Common Venue  \
4         64109  Thrift / Vintage Store                  Park   
6         64110          Sandwich Place   Fried Chicken Joint   
9         64113           Grocery Store           Coffee Shop   
10        64116          Sandwich Place         Grocery Store   
11        64127      Mexican Restaurant        Sandwich Place   
12        64118                    Park  Fast Food Restaurant   
14        64124                    Park    Mexican Restaurant   
15        64130    Fast Food Restaurant     Electronics Store   
19        64145          Sandwich Place         Big Box Store   
20        64151             Coffee Shop           Video Store   
28        64168             Flower Shop        Sandwich Place   

     3rd Most Common Venue 4th Most Common Venue  5th Most Common Venue  \
4     Fast Food Restaurant              Bus Stop         Discount Store   
6     Fast Food Restaurant              Bus Stop           Food Service   
9      American Restaurant        Sandwich Place         Ice Cream Shop   
10    Fast Food Restaurant     Convenience Store                Brewery   
11          Discount Store      Business Service   Fast Food Restaurant   
12     American Restaurant    Mexican Restaurant             Hobby Shop   
14       Convenience Store           Coffee Shop      Mobile Phone Shop   
15          Shopping Plaza              Bus Stop             Shoe Store   
19  Furniture / Home Store   Arts & Crafts Store        Automotive Shop   
20    Fast Food Restaurant        Breakfast Spot                   Park   
28             Supermarket          Soccer Field  General Entertainment   

   6th Most Common Venue 7th Most Common Venue   8th Most Common Venue  \
4                Theater             Juice Bar         Automotive Shop   
6           Liquor Store                Museum             Bus Station   
9     Mexican Restaurant           Flea Market          Shopping Plaza   
10  Outdoor Supply Store                   Pub            Intersection   
11                  Park             Pawn Shop            Intersection   
12        Sandwich Place            Restaurant      Chinese Restaurant   
14                Museum            Food Truck              Taco Place   
15        Discount Store        Sandwich Place          Baseball Field   
19  Fast Food Restaurant    Golf Driving Range          Discount Store   
20            Restaurant    Chinese Restaurant  Thrift / Vintage Store   
28         Deli / Bodega  Fast Food Restaurant     American Restaurant   

        9th Most Common Venue           10th Most Common Venue  
4                 Candy Store  Southern / Soul Food Restaurant  
6               Grocery Store                        Juice Bar  
9             Supplement Shop                      Flower Shop  
10      Street Food Gathering                   Discount Store  
11  Latin American Restaurant                  Warehouse Store  
12                        Gym                        Pawn Shop  
14  Latin American Restaurant              Fried Chicken Joint  
15              Grocery Store              American Restaurant  
19           Department Store                      Yoga Studio  
20                   Pharmacy                              Gym  
28          Convenience Store                   Gymnastics Gym

## Discussion and Conclusion

The cluster I labeled "Industrial Oldtown" seems best suited for a community coffee shop.  The antiques and other local \
venues demonstrate a high level of community value.  The area seems to support local shops over nationwide chains \
more than the average community.  The cluster contains a constant flow of potential customers because an abundance of \
industrial jobs attracts workers who live outside of the area.  More specifically, the coffee shop should look to open \
within the postal code 64109 region because it remains the only area without a coffee shop as a top 10 common venue. \
Postal code 64109 makes up the southern portion of an area known as the West Bottoms.  The combination of local support \
and less competition make the southern portion of the West Bottoms community in Kansas City an ideal place to open a \
community and relationship driven coffee shop.